docs/csharp/nullable-warnings.md

```diff
@@ -0,0 +1,130 @@
+
+## Possible dereference of null
+
+One set of warnings alert you that you're dereferencing a variable whose *null-state* is *maybe-null*. One example might be:
+
+```csharp
+string message = null;
+Console.WriteLine(message.Length);
+```
+
+To remove these warnings, you need to add code to change that variable's *null-state* to *not-null* before dereferencing it.
+
+In many instances, you can fix these warnings by checking that a variable isn't null before dereferencing it. For example, the above example could be rewritten as:
+
+:::code language="csharp" source="snippets/null-warnings/Program.cs" id="ProvideNullCheck":::
+
+When your code generates a warning that it may be dereferencing a *maybe-null* reference, make sure you've done a null check. If you haven't, add one. The compiler warning helped you address a possible bug.
+
+Other instances when you get these warnings may be false positive. You may have a private utility method that tests for null. The compiler doesn't know that the method provides a null check. Consider the following example:
+
+:::code language="csharp" source="./snippets/null-warnings/NullTests.cs" id="PrivateNullTest":::
+
+The compiler warns that you may be dereferencing null when you write the property `message.Length` because its static analysis determines that `message` may be `null`. You may know that `IsNotNull` provides a null check, and when it returns `true`, the *null-state* of `message` should be *not-null*. You must tell the compiler those facts. One way is to use the null forgiving operator, `!`. You can change the `WriteLine` statement to match the following code:
+
+```csharp
+Console.WriteLine(message!.Length);
+```
+
+The null forgiving operator makes the expression *not-null* even if it was *maybe-null* without the `!` applied. In this example, a better solution is to add an attribute to the signature of `IsNotNull`:
+
+:::code language="csharp" source="./snippets/null-warnings/NullTests.cs" id="AnnotatedNullCheck":::
+
+The <xref:System.Diagnostics.CodeAnalysis.NotNullWhenAttribute?displayProperty=nameWithType> informs the compiler that the argument used for the `obj` parameter is *not-null* when the method returns `true`. When the method returns `false`, the argument has the same *null-state* it had before the method was called.
+
+There's a rich set of attributes you can use to describe how your methods and properties affect *null-state*. You can learn about them in the language reference article on [Nullable static analysis attributes](language-reference/attributes/nullable-analysis.md).
```

Comment:
```suggestion
> [!TIP]
> There's a rich set of attributes you can use to describe how your methods and properties affect *null-state*. You can learn about them in the language reference article on [Nullable static analysis attributes](language-reference/attributes/nullable-analysis.md).
```

I suggest promoting this to a tip as it's really good callout, but it's your call.